pkg/sql/gcjob/gc_job.go

```diff
@@ -95,6 +98,70 @@ func performGC(
 	return nil
 }
 
+// performGC GCs any schema elements that are in the DELETING state and returns
+// a bool indicating if it GC'd any elements.
+func deleteData(
+	ctx context.Context,
+	execCfg *sql.ExecutorConfig,
+	details *jobspb.SchemaChangeGCDetails,
+	progress *jobspb.SchemaChangeGCProgress,
+) error {
+	switch {
+	case details.Indexes != nil:
+		return errors.Wrap(
+			deleteIndexData(ctx, execCfg, details.ParentID, progress),
+			"attempting to delete index data",
+		)
+	case details.Tables != nil:
+		return errors.Wrap(
+			deleteTableData(ctx, execCfg, progress),
+			"attempted to delete table data",
+		)
+	default:
+		return nil
+	}
+}
+
+func deleteTableData(
+	ctx context.Context, cfg *sql.ExecutorConfig, progress *jobspb.SchemaChangeGCProgress,
+) error {
+	if log.ExpensiveLogEnabled(ctx, 2) {
+		log.Infof(ctx, "GC is being considered for tables: %+v", progress.Tables)
+	}
+	for _, droppedTable := range progress.Tables {
+		var table catalog.TableDescriptor
+		if err := sql.DescsTxn(ctx, cfg, func(ctx context.Context, txn *kv.Txn, col *descs.Collection) (err error) {
+			table, err = col.Direct().MustGetTableDescByID(ctx, txn, droppedTable.ID)
+			return err
+		}); err != nil {
+			if errors.Is(err, catalog.ErrDescriptorNotFound) {
+				// This can happen if another GC job created for the same table got to
+				// the table first. See #50344.
+				log.Warningf(ctx, "table descriptor %d not found while attempting to GC, skipping", droppedTable.ID)
+				// Update the details payload to indicate that the table was dropped.
+				markTableGCed(ctx, droppedTable.ID, progress, jobspb.SchemaChangeGCProgress_CLEARED)
+				continue
+			}
+			return errors.Wrapf(err, "fetching table %d", droppedTable.ID)
+		}
+
+		// TODO(ajwerner): How does this happen?
+		if !table.Dropped() {
+			// We shouldn't drop this table yet.
+			continue
+		}
+		if err := DeleteAllTableData(ctx, cfg.DB, cfg.DistSender, cfg.Codec, table); err != nil {
```

Comment:
Would be nice to checkpoint our progress here, but given #85984 it should be fairly cheap to write these again.

Reply:
In a lot of cases the checkpoint would be more expensive than redoing the work I suspect. I'm going to not do this if you don't mind.